File: reference/src/mcp_calculator/calculator.py
"""
Core calculation engine using Python's Decimal module for precision.

This module provides the calculation logic independent of the MCP transport layer,
enabling use as a library or through the MCP server.
"""

from decimal import Decimal, ROUND_HALF_UP, InvalidOperation
from typing import Any, Dict, List, Optional, Union
import math


class CalculationEngine:
    """
    Precise arithmetic calculation engine using Python's Decimal module.

    Supports batch operations with variable references, enabling complex
    multi-step calculations in a single call.

    Example:
        engine = CalculationEngine()
        result = engine.execute([
            {"name": "oct_rate", "operation": "divide", "args": [2561276, 8]},
            {"name": "sep_rate", "operation": "divide", "args": [8782334, 21]},
            {"name": "change", "operation": "percentage", "args": ["oct_rate", "sep_rate"]}
        ])
        # Returns: {"oct_rate": 320159.5, "sep_rate": 418206.38..., "change": -23.44...}
    """

    def __init__(self, decimal_places: int = 10):
        """
        Initialize the calculation engine.

        Args:
            decimal_places: Default precision for rounding operations
        """
        self.decimal_places = decimal_places
        self._results: Dict[str, Any] = {}

    def execute(self, calculations: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Execute a batch of calculations.

        Args:
            calculations: List of calculation specs, each with:
                - name: Optional label for the result
                - operation: Operation to perform
                - args: Arguments (numbers or references to previous results)

        Returns:
            Dict with "success" flag and "results" dict of named values
        """
        self._results = {}

        for i, calc in enumerate(calculations):
            name = calc.get("name", f"result_{i}")
            operation = calc.get("operation", "").lower()
            args = calc.get("args", [])

            try:
                # Resolve variable references
                resolved_args = self._resolve_args(args)

                # Convert to Decimal for precision
                decimal_args = [Decimal(str(a)) for a in resolved_args]

                # Execute operation
                result = self._execute_operation(operation, decimal_args)

                # Store result (convert to float for JSON serialization)
                self._results[name] = float(result) if isinstance(result, Decimal) else result

            except (InvalidOperation, ValueError, ZeroDivisionError) as e:
                self._results[name] = {"error": str(e)}

        return {
            "success": True,
            "results": self._results
        }

    def _resolve_args(self, args: List[Any]) -> List[Union[int, float]]:
        """Resolve variable references in arguments."""
        resolved = []
        for arg in args:
            if isinstance(arg, str) and arg in self._results:
                value = self._results[arg]
                if isinstance(value, dict) and "error" in value:
                    raise ValueError(f"Cannot use errored result '{arg}'")
                resolved.append(value)
            else:
                resolved.append(arg)
        return resolved

    def _execute_operation(self, operation: str, args: List[Decimal]) -> Union[Decimal, bool]:
        """Execute a single operation."""

        if operation == "add":
            if len(args) < 2:
                raise ValueError("add requires at least 2 arguments")
            return sum(args)

        elif operation == "subtract":
            if len(args) < 2:
                raise ValueError("subtract requires at least 2 arguments")
            return args[0] - args[1]

        elif operation == "multiply":
            if len(args) < 2:
                raise ValueError("multiply requires at least 2 arguments")
            result = args[0]
            for arg in args[1:]:
                result *= arg
            return result

        elif operation == "divide":
            if len(args) < 2:
                raise ValueError("divide requires at least 2 arguments")
            if args[1] == 0:
                raise ValueError("division by zero")
            return args[0] / args[1]

        elif operation == "sum":
            return sum(args)

        elif operation == "average":
            if not args:
                raise ValueError("average requires at least 1 argument")
            return sum(args) / len(args)

        elif operation == "min":
            if not args:
                raise ValueError("min requires at least 1 argument")
            return min(args)

        elif operation == "max":
            if not args:
                raise ValueError("max requires at least 1 argument")
            return max(args)

        elif operation == "percentage":
            # percentage(new, old) = ((new - old) / old) * 100
            if len(args) < 2:
                raise ValueError("percentage requires 2 arguments: new, old")
            new_val, old_val = args[0], args[1]
            if old_val == 0:
                raise ValueError("cannot calculate percentage with zero base")
            return ((new_val - old_val) / old_val) * 100

        elif operation == "round":
            if not args:
                raise ValueError("round requires at least 1 argument")
            decimals = int(args[1]) if len(args) > 1 else self.decimal_places
            return args[0].quantize(
                Decimal(10) ** -decimals, rounding=ROUND_HALF_UP
            )

        elif operation == "abs":
            if not args:
                raise ValueError("abs requires 1 argument")
            return abs(args[0])

        elif operation == "ceil":
            if not args:
                raise ValueError("ceil requires 1 argument")
            return Decimal(math.ceil(args[0]))

        elif operation == "floor":
            if not args:
                raise ValueError("floor requires 1 argument")
            return Decimal(math.floor(args[0]))

        elif operation == "compare":
            if len(args) < 2:
                raise ValueError("compare requires at least 2 arguments")
            a, b = args[0], args[1]
            op = str(args[2]) if len(args) > 2 else "<"

            if op == "<":
                return a < b
            elif op == ">":
                return a > b
            elif op == "<=":
                return a <= b
            elif op == ">=":
                return a >= b
            elif op == "==" or op == "=":
                return a == b
            else:
                raise ValueError(f"unknown comparison operator: {op}")

        elif operation == "median":
            if not args:
                raise ValueError("median requires at least 1 argument")
            sorted_args = sorted(args)
            n = len(sorted_args)
            mid = n // 2
            if n % 2 == 0:
                return (sorted_args[mid - 1] + sorted_args[mid]) / 2
            return sorted_args[mid]

        elif operation == "stddev":
            if len(args) < 2:
                raise ValueError("stddev requires at least 2 arguments")
            mean = sum(args) / len(args)
            variance = sum((x - mean) ** 2 for x in args) / len(args)
            return Decimal(str(math.sqrt(float(variance))))

        elif operation == "roi":
            # roi(gain, cost) = (gain - cost) / cost * 100
            if len(args) < 2:
                raise ValueError("roi requires 2 arguments: gain, cost")
            gain, cost = args[0], args[1]
            if cost == 0:
                raise ValueError("cannot calculate ROI with zero cost")
            return ((gain - cost) / cost) * 100

        elif operation == "compound_interest":
            # compound_interest(principal, rate, periods) = principal * (1 + rate) ^ periods
            if len(args) < 3:
                raise ValueError("compound_interest requires 3 arguments: principal, rate, periods")
            principal, rate, periods = args[0], args[1], int(args[2])
            return principal * ((1 + rate) ** periods)

        elif operation == "present_value":
            # present_value(future_value, rate, periods) = fv / (1 + rate) ^ periods
            if len(args) < 3:
                raise ValueError("present_value requires 3 arguments: future_value, rate, periods")
            fv, rate, periods = args[0], args[1], int(args[2])
            return fv / ((1 + rate) ** periods)

        else:
            raise ValueError(f"unknown operation: {operation}")


def calculate(calculations: List[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Convenience function to perform calculations without instantiating an engine.

    Args:
        calculations: List of calculation specs

    Returns:
        Dict with "success" flag and "results" dict

    Example:
        result = calculate([
            {"name": "total", "operation": "sum", "args": [100, 200, 300]},
            {"name": "avg", "operation": "average", "args": [100, 200, 300]}
        ])
        # Returns: {"success": True, "results": {"total": 600, "avg": 200}}
    """
    engine = CalculationEngine()
    return engine.execute(calculations)

File: internal/calculator/calculator.go
// Package calculator provides precise arithmetic calculations using decimal arithmetic.
// It eliminates floating-point errors common in IEEE 754 arithmetic.
package calculator

import (
	"fmt"
	"math"
	"sort"

	"github.com/shopspring/decimal"
)

// Calculation represents a single calculation to perform.
type Calculation struct {
	Name      string `json:"name,omitempty"`
	Operation string `json:"operation"`
	Args      []any  `json:"args"`
}

// Result represents the result of a batch calculation.
type Result struct {
	Success bool           `json:"success"`
	Results map[string]any `json:"results"`
}

// Engine performs precise arithmetic calculations.
type Engine struct {
	decimalPlaces int32
	results       map[string]any
}

// NewEngine creates a new calculation engine with default precision (10 decimal places).
func NewEngine() *Engine {
	return &Engine{
		decimalPlaces: 10,
		results:       make(map[string]any),
	}
}

// NewEngineWithPrecision creates a new calculation engine with custom precision.
func NewEngineWithPrecision(places int32) *Engine {
	return &Engine{
		decimalPlaces: places,
		results:       make(map[string]any),
	}
}

// Execute performs a batch of calculations.
func (e *Engine) Execute(calculations []Calculation) Result {
	e.results = make(map[string]any)

	for i, calc := range calculations {
		name := calc.Name
		if name == "" {
			name = fmt.Sprintf("result_%d", i)
		}

		result, err := e.executeOne(calc.Operation, calc.Args)
		if err != nil {
			e.results[name] = map[string]any{"error": err.Error()}
		} else {
			e.results[name] = result
		}
	}

	return Result{
		Success: true,
		Results: e.results,
	}
}

// resolveArg resolves a single argument, which may be a number or a reference.
// The visited map tracks already-resolved variable names to detect circular references.
func (e *Engine) resolveArg(arg any) (decimal.Decimal, error) {
	return e.resolveArgWithVisited(arg, make(map[string]bool))
}

// resolveArgWithVisited resolves an argument with cycle detection.
func (e *Engine) resolveArgWithVisited(arg any, visited map[string]bool) (decimal.Decimal, error) {
	switch v := arg.(type) {
	case string:
		// Check for circular reference
		if visited[v] {
			return decimal.Zero, fmt.Errorf("circular reference detected: '%s'", v)
		}

		// Variable reference
		if result, ok := e.results[v]; ok {
			if errMap, isErr := result.(map[string]any); isErr {
				if _, hasError := errMap["error"]; hasError {
					return decimal.Zero, fmt.Errorf("cannot use errored result '%s'", v)
				}
			}
			// Mark as visited before recursing
			visited[v] = true
			return e.resolveArgWithVisited(result, visited)
		}
		// Try to parse as number string
		return decimal.NewFromString(v)
	case int:
		return decimal.NewFromInt(int64(v)), nil
	case int32:
		return decimal.NewFromInt(int64(v)), nil
	case int64:
		return decimal.NewFromInt(v), nil
	case float32:
		return decimal.NewFromFloat32(v), nil
	case float64:
		return decimal.NewFromFloat(v), nil
	case decimal.Decimal:
		return v, nil
	default:
		return decimal.Zero, fmt.Errorf("unsupported argument type: %T", arg)
	}
}

// resolveArgs resolves all arguments in a list.
func (e *Engine) resolveArgs(args []any) ([]decimal.Decimal, error) {
	resolved := make([]decimal.Decimal, len(args))
	for i, arg := range args {
		val, err := e.resolveArg(arg)
		if err != nil {
			return nil, err
		}
		resolved[i] = val
	}
	return resolved, nil
}

// executeOne executes a single operation.
func (e *Engine) executeOne(operation string, args []any) (any, error) {
	// Special handling for compare - only resolve first 2 args as decimals
	if operation == "compare" {
		return e.opCompare(args)
	}

	resolved, err := e.resolveArgs(args)
	if err != nil {
		return nil, err
	}

	switch operation {
	case "add":
		return e.opAdd(resolved)
	case "subtract":
		return e.opSubtract(resolved)
	case "multiply":
		return e.opMultiply(resolved)
	case "divide":
		return e.opDivide(resolved)
	case "sum":
		return e.opSum(resolved)
	case "average":
		return e.opAverage(resolved)
	case "min":
		return e.opMin(resolved)
	case "max":
		return e.opMax(resolved)
	case "median":
		return e.opMedian(resolved)
	case "stddev":
		return e.opStddev(resolved)
	case "percentage":
		return e.opPercentage(resolved)
	case "roi":
		return e.opROI(resolved)
	case "compound_interest":
		return e.opCompoundInterest(resolved)
	case "present_value":
		return e.opPresentValue(resolved)
	case "round":
		return e.opRound(resolved, args)
	case "abs":
		return e.opAbs(resolved)
	case "ceil":
		return e.opCeil(resolved)
	case "floor":
		return e.opFloor(resolved)
	default:
		return nil, fmt.Errorf("unknown operation: %s", operation)
	}
}

// decimalToNumber converts a Decimal to a numeric value.
// For integers, returns int64 to avoid floating point representation issues.
// For decimals, returns the float64 representation.
func decimalToNumber(d decimal.Decimal) any {
	// If it's an integer, return as int64 for cleaner JSON output
	if d.Equal(d.Truncate(0)) {
		return d.IntPart()
	}
	// Otherwise return as float64
	f, _ := d.Float64()
	return f
}

func (e *Engine) opAdd(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("add requires at least 2 arguments")
	}
	result := args[0]
	for _, arg := range args[1:] {
		result = result.Add(arg)
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opSubtract(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("subtract requires at least 2 arguments")
	}
	// Variadic: subtract every arg after the first from the running result.
	// Matches add/multiply's left-fold semantics so subtract(a, b, c) is a-b-c.
	result := args[0]
	for _, arg := range args[1:] {
		result = result.Sub(arg)
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opMultiply(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("multiply requires at least 2 arguments")
	}
	result := args[0]
	for _, arg := range args[1:] {
		result = result.Mul(arg)
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opDivide(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("divide requires at least 2 arguments")
	}
	// Variadic: divide by every arg after the first (left-fold). Matches
	// the symmetry with subtract's variadic semantics.
	result := args[0]
	for _, arg := range args[1:] {
		if arg.IsZero() {
			return nil, fmt.Errorf("division by zero")
		}
		result = result.Div(arg)
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opSum(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return int64(0), nil
	}
	result := decimal.Zero
	for _, arg := range args {
		result = result.Add(arg)
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opAverage(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("average requires at least 1 argument")
	}
	sum := decimal.Zero
	for _, arg := range args {
		sum = sum.Add(arg)
	}
	result := sum.Div(decimal.NewFromInt(int64(len(args))))
	return decimalToNumber(result), nil
}

func (e *Engine) opMin(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("min requires at least 1 argument")
	}
	result := args[0]
	for _, arg := range args[1:] {
		if arg.LessThan(result) {
			result = arg
		}
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opMax(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("max requires at least 1 argument")
	}
	result := args[0]
	for _, arg := range args[1:] {
		if arg.GreaterThan(result) {
			result = arg
		}
	}
	return decimalToNumber(result), nil
}

func (e *Engine) opMedian(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("median requires at least 1 argument")
	}

	// Sort a copy
	sorted := make([]decimal.Decimal, len(args))
	copy(sorted, args)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].LessThan(sorted[j])
	})

	n := len(sorted)
	mid := n / 2

	var result decimal.Decimal
	if n%2 == 0 {
		result = sorted[mid-1].Add(sorted[mid]).Div(decimal.NewFromInt(2))
	} else {
		result = sorted[mid]
	}

	return decimalToNumber(result), nil
}

func (e *Engine) opStddev(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("stddev requires at least 2 arguments")
	}

	// Calculate mean
	sum := decimal.Zero
	for _, arg := range args {
		sum = sum.Add(arg)
	}
	mean := sum.Div(decimal.NewFromInt(int64(len(args))))

	// Calculate variance
	variance := decimal.Zero
	for _, arg := range args {
		diff := arg.Sub(mean)
		variance = variance.Add(diff.Mul(diff))
	}
	variance = variance.Div(decimal.NewFromInt(int64(len(args))))

	// Calculate stddev (sqrt of variance)
	varianceFloat, _ := variance.Float64()
	return math.Sqrt(varianceFloat), nil
}

func (e *Engine) opPercentage(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("percentage requires 2 arguments: new, old")
	}
	newVal, oldVal := args[0], args[1]
	if oldVal.IsZero() {
		return nil, fmt.Errorf("cannot calculate percentage with zero base")
	}
	// ((new - old) / old) * 100
	result := newVal.Sub(oldVal).Div(oldVal).Mul(decimal.NewFromInt(100))
	return decimalToNumber(result), nil
}

func (e *Engine) opROI(args []decimal.Decimal) (any, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("roi requires 2 arguments: gain, cost")
	}
	gain, cost := args[0], args[1]
	if cost.IsZero() {
		return nil, fmt.Errorf("cannot calculate ROI with zero cost")
	}
	// ((gain - cost) / cost) * 100
	result := gain.Sub(cost).Div(cost).Mul(decimal.NewFromInt(100))
	return decimalToNumber(result), nil
}

func (e *Engine) opCompoundInterest(args []decimal.Decimal) (any, error) {
	if len(args) < 3 {
		return nil, fmt.Errorf("compound_interest requires 3 arguments: principal, rate, periods")
	}
	principal, rate := args[0], args[1]
	periods, _ := args[2].Float64()

	// principal * (1 + rate)^periods
	onePlusRate := decimal.NewFromInt(1).Add(rate)
	onePlusRateFloat, _ := onePlusRate.Float64()
	multiplier := math.Pow(onePlusRateFloat, periods)
	principalFloat, _ := principal.Float64()

	return principalFloat * multiplier, nil
}

func (e *Engine) opPresentValue(args []decimal.Decimal) (any, error) {
	if len(args) < 3 {
		return nil, fmt.Errorf("present_value requires 3 arguments: future_value, rate, periods")
	}
	fv, rate := args[0], args[1]
	periods, _ := args[2].Float64()

	// fv / (1 + rate)^periods
	onePlusRate := decimal.NewFromInt(1).Add(rate)
	onePlusRateFloat, _ := onePlusRate.Float64()
	divisor := math.Pow(onePlusRateFloat, periods)
	fvFloat, _ := fv.Float64()

	return fvFloat / divisor, nil
}

func (e *Engine) opRound(args []decimal.Decimal, rawArgs []any) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("round requires at least 1 argument")
	}

	places := e.decimalPlaces
	if len(rawArgs) > 1 {
		if p, err := e.resolveArg(rawArgs[1]); err == nil {
			places = int32(p.IntPart())
		}
	}

	result := args[0].Round(places)
	return decimalToNumber(result), nil
}

func (e *Engine) opAbs(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("abs requires 1 argument")
	}
	result := args[0].Abs()
	return decimalToNumber(result), nil
}

func (e *Engine) opCeil(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("ceil requires 1 argument")
	}
	result := args[0].Ceil()
	return decimalToNumber(result), nil
}

func (e *Engine) opFloor(args []decimal.Decimal) (any, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("floor requires 1 argument")
	}
	result := args[0].Truncate(0)
	return decimalToNumber(result), nil
}

func (e *Engine) opCompare(rawArgs []any) (bool, error) {
	if len(rawArgs) < 2 {
		return false, fmt.Errorf("compare requires at least 2 arguments")
	}

	// Resolve only the first two arguments as decimals
	a, err := e.resolveArg(rawArgs[0])
	if err != nil {
		return false, err
	}
	b, err := e.resolveArg(rawArgs[1])
	if err != nil {
		return false, err
	}

	op := "<"
	if len(rawArgs) > 2 {
		if opStr, ok := rawArgs[2].(string); ok {
			op = opStr
		}
	}

	switch op {
	case "<":
		return a.LessThan(b), nil
	case ">":
		return a.GreaterThan(b), nil
	case "<=":
		return a.LessThanOrEqual(b), nil
	case ">=":
		return a.GreaterThanOrEqual(b), nil
	case "==", "=":
		return a.Equal(b), nil
	default:
		return false, fmt.Errorf("unknown comparison operator: %s", op)
	}
}

// Calculate is a convenience function to perform calculations without instantiating an engine.
func Calculate(calculations []Calculation) Result {
	engine := NewEngine()
	return engine.Execute(calculations)
}
